tools/generator/dfg/rp/rp_identifier.py

```diff
@@ -0,0 +1,41 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2013     , Kevin Läufer
+# Copyright (c) 2013-2014, Niklas Hauser
+# Copyright (c)      2016, Fabian Greif
+# Copyright (c)      2020, Hannes Ellinger
+# All rights reserved.
+
+import re
+import logging
+
+from modm_devices.device_identifier import DeviceIdentifier
+
+LOGGER = logging.getLogger("dfg.rp.identifier")
+
+class RPIdentifier:
+    """ RPIdentifier
+    A class to parse RP device strings, e.g. RP2040.
+    Device names are organized as follows:
+          RP      2      0    4    0
+      {platform}{cores}{cpu}{ram}{flash}
+    """
+
+    @staticmethod
+    def from_string(string):
+        string = string.lower()
+
+        if string.startswith("rp"):
+            i = DeviceIdentifier("{platform}{cores}{cpu}{ram}{flash}")
+            i.set("platform", "rp")
+            i.set("cores", string[2])
+            i.set("cpu", string[3])
+            i.set("ram", string[4])
+            i.set("flash", string[5])
+            i.set("name", string)
+            i.set("family", string[2:])
```

Comment:
I'm not sure, but I think we only need `platform` and `family` and `name` are not required, they are just a STM32 specific identifier thing.
Perhaps the family could just be `20`, since the RAM and Flash differences are probably not relevant to the peripheral differences.

![](https://user-images.githubusercontent.com/163066/148317789-a4bb504d-7d92-41b6-8cb4-23b8157fa142.png)


I would recommend `{platform}{cores}{type}{ram}{flash}` so that `cpu` can be reserved for the CPU ID like so: `{platform}{cores}{type}{ram}{flash}@{cpu}` with `rp2040@0` and `rp2040@1` idenfitying which CPU your application targets.
This could be useful for things like chosing the RAM location for the main stack etc, although I'm not entirely sure how to properly support dual-core in SCons yet (since it would need to combine two firmwares into a single binary).

This is similiar to the [`stm32h755zit7@m4` and `stm32h755zit7@m7`](https://github.com/modm-io/modm-devices/blob/develop/devices/stm32/stm32h7-45_55.xml) idenfiers for the dual-core STM32 devices. The [STM32 DFG duplicates the device](https://github.com/modm-io/modm-devices/blob/develop/tools/generator/dfg/stm32/stm_device_tree.py#L68-L81) for the cores and then merges them again later.

Reply:
in difference to stm32h755zit rp2040 is SMP system and dnt need separate compile for cores. Cores mostly equivalent and share all resources. esp32 more similar to this processor. We can see the implementation working with multiple cores at esp-idf and their port of FreeRTOS. I think for first need a simple interface like  `run_on_core1(func)`. And later adapt rtos and fibers.

Reply:
Ah, that is a very good point! I only skimmed the datasheet and peeked into the Pico SKD, so I wasn't sure how they solved it, but it makes a lot of sense to use one firmware with SMP like this.

I would still recommend using `{platform}{cores}{type}{ram}{flash}` to avoid confusion with the STM32 AMP `cpu` and also the core driver attribute "type". Let's also remove `name` and turn `family` into `20`.